src/vsr/replica.zig

```diff
@@ -4899,112 +4948,322 @@ pub fn ReplicaType(
+            //         4  (4 5 6 7 8)          8 (didn't participate in view change)
+            //
+            // where the new leader's headers depends on which of replica 1 and 2's DVCs is used
+            // for repair before the other (i.e. whether they repair op 6 or 7 first).
+            //
+            // For the above case to occur, replicas 0, 1, and 2 must all share the highest `view_normal`.
+            // And since they share the latest `view_normal`, ops 5,6,7 were just installed by
+            // `replace_header`, which is order-independent (it doesn't use the hash chain).
+            //
+            // (If replica 0's view_normal was greater than 1/2's, then replica 0 must have all
+            // headers from previous views. Which means 6,7 are from the current view. But since
+            // replica 0 doesn't have 6/7, then replica 1/2 must share the latest view_normal. ∎)
+
+            // Now that the canonical headers are all in place, repair any other headers:
+            for (self.do_view_change_from_all_replicas) |received| {
+                if (received) |message| {
+                    const view_normal = @intCast(u32, message.header.timestamp);
+                    assert(view_normal < message.header.view);
+
+                    if (view_normal == view_normal_canonical) continue;
+                    assert(view_normal < view_normal_canonical);
+
+                    for (message_body_as_headers(message)) |*header| {
+                        // We must trust headers that other replicas have committed, because
+                        // repair_header() will not repair a header if the hash chain has a gap.
+                        if (header.op <= message.header.commit) {
+                            log.debug(
+                                "{}: on_do_view_change: committed: replica={} op={} checksum={}",
+                                .{
+                                    self.replica,
+                                    message.header.replica,
+                                    header.op,
+                                    header.checksum,
+                                },
+                            );
+                            self.replace_header(header);
+                        } else {
+                            _ = self.repair_header(header);
+                        }
+                    }
+                }
+            }
+
+            // The `prepare_timestamp` prevents a primary's own clock from running backwards.
+            // Therefore, `prepare_timestamp`:
+            // 1. is advanced if behind the cluster, but never reset if ahead of the cluster, i.e.
+            // 2. may not always reflect the timestamp of the latest prepared op, and
+            // 3. should be advanced before discarding the timestamps of any uncommitted headers.
+            const head = self.journal.header_with_op(self.op).?;
```

Comment:
```suggestion
            const head_timestamp = self.journal.header_with_op(self.op).?.timestamp;
```

To shrink the scope further for purpose.

Reply:
We could also consider adding `timestamp` to `do_view_change_quorum_head` since that is what we are thinking about here... does any quorum have a newer prepare timestamp?